perun/utils/common/common_kit.py

```diff
@@ -437,4 +451,97 @@ def get_module(module_name: str) -> types.ModuleType:
     return MODULE_CACHE[module_name]
 
 
+def compact_convert_list_to_str(
+    number_list: list[int | float] | array.array[float] | array.array[int], float_precision: int = 2
+) -> list[str]:
+    """Converts list to list of compact strings
+
+    :param number_list: list of numbers
+    :param float_precision: float precision of the numbers
+    :return list of compact strings
+    """
+    return [compact_convert_num_to_str(num, 2) for num in number_list]
+
+
+def compact_convert_num_to_str(number: int | float, float_precision: int = 2) -> str:
+    """Converts the number to the smallest string possible
+
+    :param number: converted number
+    :param float_precision: float precision, i.e. number of decimal places in number
+    :return compact string
+    """
+    return str(to_compact_num(number, float_precision))
+
+
+def to_compact_num(number: int | float, float_precision: int = 2) -> int | float:
+    """Converts the number to the smallest string possible
+
+    :param number: converted number
+    :param float_precision: float precision, i.e. number of decimal places in number
+    :return compact num
+    """
+    if isinstance(number, int):
+        return number
+    elif number.is_integer():
+        return int(number)
+    else:
+        return round(number, float_precision)
+
+
+@contextlib.contextmanager
+def disposable_resources(disposable: Any) -> Any:
+    """Helper context manager that disposes of the passed object
+
+    :param disposable: object that should be disposed immediately
+    :return: object
+    """
+    try:
+        yield disposable
+    except Exception:
+        # Re-raise the encountered exception
+        raise
+    finally:
+        del disposable
+        gc.collect()
+
+
+def binary_search(
+    values: list[Any], value: Any, low: int, high: int, key: Callable[[Any], Any] = lambda x: x
+) -> int:
+    """Classical binary search algorithm
+
+    :param values: list of any values
+    :param value: value for which we are looking up the insertion point
+    :param key: key used for getting the key
+    :param low: left pivot
+    :param high: right pivot
+    :return: insertion point for the value in values
+    """
+    while low <= high:
+        mid = low + (high - low) // 2
+        if value == values[mid]:
+            return mid + 1
+        elif value > values[mid]:
+            low = mid + 1
+        else:
+            high = mid - 1
+    return low
+
+
+def add_to_sorted(
+    values: list[Any], value: Any, key: Callable[[Any], Any] = lambda x: x, max_pick: int = 0
+) -> None:
```

Comment:
Same, the bisect module should be able to do this as well?

Reply:
I don't think I can fully replace this function, since I want to keep only `max_pick` elements in the list, however, it could be extracted to higher function. Though I will replace the binary search + insertion with insort (unless it breaks something)